lib/fastly/client.rb

```diff
@@ -129,8 +164,19 @@ def post_and_put(method, path, params = {})
       extras = params.delete(:headers) || {}
       include_auth = params.key?(:include_auth) ? params.delete(:include_auth) : true
       query = make_params(params)
-      resp  = http.send(method, path, query, headers(extras, include_auth).merge('Content-Type' =>  'application/x-www-form-urlencoded'))
-      fail Error, resp.body unless resp.kind_of?(Net::HTTPSuccess)
+      tries = 0
+      begin
+        resp  = http.send(method, path, query, headers(extras, include_auth).merge('Content-Type' =>  'application/x-www-form-urlencoded'))
+        fail Error, resp.body unless resp.kind_of?(Net::HTTPSuccess)
+      rescue ::Fastly::Error => e
+        if should_retry(tries, e.message)
+          tries += 1
+          sleep(RETRY_TIME)
+          retry
+        else
+          raise
+        end
+      end
```

Comment:
There feels like a lot of duplication with this code. Can this be refactored into a single function? 

**NOTE**: I've not written Ruby in years so I'll defer to my Ruby colleagues here, but I'm thinking a function that accepts a lambda. The lambda would wrap the execution logic (e.g. `http.send`, `http.request`, `http.delete` etc and the `fail` check) and then the function accepting the lambda would contain everything else (i.e. `begin/rescue`).

Reply:
It feels like duplication to me too. I kept it simple to start with. Let me see if I can figure out a way to wrap it into a function.